### ISSUE TITLE
IPX: File not found when using assets directory

### ISSUE THREAD

**Issue by alex-lmt** [nuxt/image#1006]
I'm using Nuxt 3.7.3, Nuxt image 1.0.0-rc.2

In a component, providing image from assets directory as src prop of <NuxtImg> does not render image, I see "404 IPX: File not found" error in JS console. 
Same image with <img> tag works correctly.

This works:
`<img src="~/assets/img/banners/front-page-banner.png" alt="Banner" />`

This does not work:
`<NuxtImg src="~/assets/img/banners/front-page-banner.png" alt="Banner" />`

Import also does not work. This does not work:
```
<script lang="ts" setup>
import banner from '~/assets/img/banners/front-page-banner.png';
</script>
<template>
  <img :src="banner" alt="Banner" />
</template>
```

According documentation https://image.nuxt.com/get-started/configuration#dir I've added assets directory to nuxt.config:
`dir: 'assets/img'`

Now image is rendered normally using it this way:
`<NuxtImg src="banners/front-page-banner.png" alt="Banner" />`

But only on local development server, after building and deploying it image is not rendered and same error appears in JS console.

**Comment by madebyfabian** [comment]
So this would be the correct way after you set the `dir` config:
`<NuxtImg src="banners/front-page-banner.png" alt="Banner" />`

Where do you deploy? According to the [docs](https://image.nuxt.com/get-started/configuration#dir), other directories than `public` don't work with e.g. vercel.

> For some providers (like vercel), using a directory other than public/ for assets is not supported since resizing happens at runtime (instead of build/generate time) and source fetched from the public/ directory (deployment URL)

**Comment by alex-lmt** [comment]
> Where do you deploy? According to the [docs](https://image.nuxt.com/get-started/configuration#dir), other directories than public don't work with e.g. vercel.

I deploy to remote server in our organization intranet using CI/CD. I'm not using vercel, I did not set provider, so assume default image processing provider is used - IPX

**Comment by madebyfabian** [comment]
@alex-lmt Alright. If you don't set the `dir` config and store your images in the `/public` folder, does it work in production? Just to confirm that IPX is generally working in your production environment.

**Comment by madebyfabian** [comment]
Ah and do you deploy with `nuxt build` or `nuxt generate`?

**Comment by alex-lmt** [comment]
@madebyfabian 

>  If you don't set the dir config and store your images in the /public folder, does it work in production? Just to confirm that IPX is generally working in your production environment.
Yes. For example image from /public/img/image.png this way is working normally:
`<NuxtImg src="/img/image.png" />`

> Ah and do you deploy with nuxt build or nuxt generate?
We are using nuxt build

**Comment by alex-lmt** [comment]
> Do you also deploy the assets/img folder?

@madebyfabian I'm relatively new in Nuxt, especially to CI/CD stuff, how to add desired directory to deploy or check whether it is being delpoyed or not?

**Comment by madebyfabian** [comment]
@alex-lmt Is there a reason you want to use the assets folder? Why not just go with the public folder?

**Comment by alex-lmt** [comment]
@madebyfabian according documentation and general advices like https://masteringnuxt.com/blog/handling-assets-in-nuxt-3, I choose assets folder as I assume image may change relatively often and I don't care of physical file name (it is ok to process it during build)

**Comment by gijsdb** [comment]
I am also experiencing this issue.

**Comment by jieceng** [comment]
I am also experiencing this issue.

following.

**Comment by l0ggik** [comment]
Have the same issue.

Would be great to know how to fix this.

**Comment by alex-lmt** [comment]
Hey, people, issue was reported with Nuxt image version 1.0.0-rc.2, it's already 1.1.0 released. Does anyone tried is the issue still persists in newest version?

**Comment by l0ggik** [comment]
Yes i tried it with Version 1.1.0 and it doesn't  work.

Could be easyly reproduced locally.

In my case i put all the images in assets/images, set provider to ipx and dir to assets/images in the config.

Def works but as soon as i try build the images are missing. 

Also the bild bundle size indicate that the images are processed but the mapping seems to not be correct when build.

**Comment by blowsie** [comment]
I too have this issue

The images are in the `.output/public` dir

![image](https://github.com/nuxt/image/assets/308572/97c6a610-ac86-4238-b598-18f6c6af7388)
```
  image: {
    dir: 'public/images',
  }
```
But IPX throws 404 when using `build`, but works fine with `dev`

**Comment by blowsie** [comment]
It seems it only breaks when using a custom dir, even if its a subfolder of public

**Comment by l0ggik** [comment]
Found a workaround: Do not use Nuxt Image, use Vuetifys v-img instead.

**Comment by notflip** [comment]
I'm getting this error, even when using the default `/img/` folder. Hope this will be fixed soon, I removed the <nuxt-img elements for now.

**Comment by fabianwohlfart** [comment]
I am also off nuxt-image completely and using my own next endpoints for img generation…

**Comment by Anil-Toxy** [comment]
I am also experiencing this issue.
any workaround ?

**Comment by steel89ita** [comment]
I'm having the same issue now.
I reverted back to nuxt and modules to the last working version, but still not working.
I even tried cleaning the node_modules, .nuxt, .output folders and yarn.lock, then running npx nuxi cleanup
but still no change.

In developer console I see that the image is pointing to "http://localhost:3000/_ipx/_/icon.png"
Using the default public dir.

**Comment by fedeanimation** [comment]
Same issue here, just update all with pnpm update.

**Comment by PatoGordo** [comment]
same issue in the latest version of Nuxt and Nuxt Image

**Comment by fisher60** [comment]
I also seem to have encountered the same issue in dev.

Steps to reproduce:
1. Use `assets` directory (`assets/img` in my case), images load via regular img tag
2. Install Nuxt image as per install instructions, and configure assets/img as dir, use <NuxtImg>
3. Run app, images no longer load with `IPX_FILE_NOT_FOUND`

**Comment by blowsie** [comment]
For most of you here, Id suggest you move your images to the root of the `public` dir and retest. That worked for me.

**Comment by jyagamino** [comment]
Hi.
While this is not a fundamental solution, when I started the server on port 3001, the images were displayed correctly. Could you please try restarting the app with the following configuration in your `nuxt.config.ts`?

```
export default defineNuxtConfig({
  // Other configurations...
  devServer: {
    port: 3001,
  }
  // Other configurations...
});

```

**Comment by ztxone** [comment]
Same 404 error with /_ipx/ folder after "build" project. 
Couldn't find the solution - have to rid all NuxtImg links (

**Comment by Dino-Kupinic** [comment]
This issue only seems to happen on Windows. On MacOS `NuxtImg` works as it should

**Comment by ztxone** [comment]
> On MacOS `NuxtImg` works as it should

nope. I'm on latest macOs Sonoma

**Comment by jyagamino** [comment]
If you want to run the application properly, changing the version of `@nuxt/image` from `v1.3.0` to `v1.2.0` will make the images appear.

**Comment by iamMAHAM** [comment]
@jyagamino the images appear in development server but not in production build

**Comment by iamMAHAM** [comment]
Unfortunately, for the time being, I've had to use a classic image tag. But given that our site has over a hundred images, changing them manually would be really tiring. While waiting for this bug to be fixed, I proceeded as follows after research :
inside composables/use-asset.ts 
```typescript
export default function useAsset(path: string): string {
  const assets = import.meta.glob('~/assets/**/*', {
    eager: true,
    import: 'default',
  })

 // all my file startsWith '/' ex: /home/hero.png  you'll need to adjust
  return assets[`/assets${path}`]
}
```

then i create  inside components/NuxtImg.vue (override @nuxt/img NuxtImg component)
```vue
<script setup lang="ts">
const props = defineProps(['src', 'width', 'height', 'alt', 'class'])
const img = (props.src.startsWith('https://')
  ?? props.src.startsWith('http://')) // link from external source
  ? props.src // use external source
  : useAsset(props.src) // resolve path
</script>

<template>
  <img
    :src="img"
    :width="props.width"
    :height="props.height"
    :alt="props.alt"
    :class="props.class"
  >
</template>
```
`The images will not be optimized as they should be, but until the bug is fixed this may help. `

**Comment by Pamavoc** [comment]
Following too.

**Comment by cfdxkk** [comment]
same issue in vercel #1282

**Comment by tarreislam** [comment]
ofc it took 3 hours to find this thread.

Yeah ill just switch to v-img and lazyload

**Comment by tarreislam** [comment]
How is it that everything always suck when its build in javascript

**Comment by davestewart** [comment]
I'm attempting to solve a similar problem in Nuxt Content Assets.

I think there's a hack which could work, and that is to put your assets in a separate Nuxt layer, and point Nuxt Image at the assets folder for that layer.

In the Nuxt Image `ipxSetup()` function, it will then see the layer, and add the assets folder to its public array.

https://github.com/nuxt/image/blob/main/src/ipx.ts#L31

**Comment by hdwebpros** [comment]
I am having the same issue. I tried a Stackblitz: https://stackblitz.com/edit/nuxt-starter-ee1s9x

But I can't upload images without a paid plan.

Remote images work fine. `<NuxtImg ... />` flat out refuses to work with any local image.

I've tried the images in the `public` directory as well as `assets/images/`. 

How to reproduce:
Initialize a new Nuxt project
Install and load NuxtImg module

Then, [follow their documentation](https://image.nuxt.com/usage/nuxt-img):

---
[Usage](https://image.nuxt.com/usage/nuxt-img#usage)
<NuxtImg> outputs a native img tag directly (without any wrapper around it). Use it like you would use the <img> tag:

`<NuxtImg src="/nuxt-icon.png" />`
Will result in:

`<img src="/nuxt-icon.png" />`
With [default provider](https://image.nuxt.com/get-started/providers), you should put `/nuxt-icon.png` inside `public/` directory for Nuxt 3 make the above example work.

---

This just doesn't work at all.

**Comment by RyanClementsHax** [comment]
Is anyone able to reproduce a reproduction repo? I tried to, but I was able to get it to work.

https://codesandbox.io/p/devbox/nuxt-image-1006-ipx-public-64qh3f

**Comment by francesco-buscicchio** [comment]
Is there any news?

**Comment by FdvTerje** [comment]
> This issue only seems to happen on Windows. On MacOS `NuxtImg` works as it should

This happens on linux too.

**Comment by Yo-2** [comment]
👇In my case, this code works after running "npm run dev", but it doesn't work when using "npm run build".
```js
// nuxt.config.js
export default defineNuxtConfig({
  image: {
    dir: 'assets/images'
})
```
```vue
<template>
   <NuxtImg src="/banner.jpeg" />
</template>
```
👇This code works after running "npm run build", but it doesn't work when using "npm run dev".
```vue
<script>
import banner from "~/assets/images/banner.jpeg"
</script>
<template>
  <NuxtImg :src="banner" />
</template>
```

**Comment by FdvTerje** [comment]
I saw that the 500 error i got with these had this error: 
   "message": "[500] [IPX_ERROR] \nSomething went wrong installing the \"sharp\" module\n\nModule did not self-register: '/var/www/html/myproject/node_modules/sharp/build/Release/sharp-linux-x64.node'.
   
   ChatGPT recommended me to delete node_modules and do this: 
   
`npm install --platform=linux --arch=x64 --ignore-scripts=false --foreground-scripts --verbose sharp`

So far I have not gotten the same error again, but not sure if this is a correct fix

**Comment by Smef** [comment]
For some of the comments in here:

The assets get tree-shaken out when running build, since they're not detected as being in use at build time.

If you run generate I the URLs crawled and the IPX-rendered images are included in your `/dist`.

For a build for running a server I don't think they make it in normally, and so they aren't there when NuxtImage and IPX looks for them to render them on-the-fly. 

This is why it works in dev, but not in a build, and does work for SSG sites.

**Comment by metkm** [comment]
Not really familiar with the codebase but I would think this is the reason
https://github.com/nuxt/image/blob/79d32da6b3aaedaf5c82ef43976f3af9286e29d7/src/ipx.ts#L46-L49

I can't figure out what it the reason for the if check here, if It's not dev it defaults to public folder?

**Comment by Smef** [comment]
It's for the reason I described in my previous post. 

Here's my understanding, from discussions I've had with people about this:

IPX generates the images upon request when running in production. They are not pre-generated at build time. Nuxt is actually sitting there, listening for requests of images of a specific format, finds the original file, and then it generates the image in response to the request at the time the request is made. When the request is received it looks for the requested source file at the given path and tries to generate the optimized image from the source.

The assets directory is not directly copied into the production bundle when generate is run. References to the assets directory don't have any meaning when the code is run in production. Assets get bundled, minified, and repackaged , and re-referenced in the JS by Vite as part of the build process. Vite drops anything which isn't used in the original code, and a path for NuxtImage is just a string rather than a real import like a package. The public directory IS copied directly into the production bundle, and so image files in there can still be found by Nuxt and the IPX process when those paths are referenced.

You can import static assets to be versioned with Vite, but my understanding is that since IPX is never actually called to generate the images, the images/references/imports are considered "unused" and get dropped by Vite. Not ideal, but that's the explanation which I've been given.

**Comment by metkm** [comment]
@Smef Yeah I think you're right, from the tests I've made removing that ternary check and just using `publicDirs` works on both development and production. It didn't seem like broke anything else.

**Comment by Smef** [comment]
Also, keep in mind that those image paths can actually be from completely separate domains. The images don't even need to be on the same server, or even the same domain, and can actually be at a completely different address! IPX just checks whatever path at runtime, gets the content, and does the processing then.

**Comment by DanilMiroshin** [comment]
From nuxt image documentation:
`Image stored in the assets/ directory are not processed with Nuxt Image because those images are managed by your bundler (such as Vite or webpack).`
https://image.nuxt.com/get-started/providers

**Comment by heydayle** [comment]
Does anyone have a solution to this problem?

**Comment by lafiefp** [comment]
For me the solution was to put the dir-config under image.ipx.fs.dir .

```diff
 export default defineNuxtConfig({
     ...
     image: {
-      dir: process.env.NUX_IMAGES_ROOT,
+      provider: "ipx",
+      ipx: {
+        fs: { dir: process.env.NUXT_IMAGES_ROOT },
+      },
     },
     ...
   });
```

**Comment by bonomiandreia** [comment]
hello!
For anyone with the same problem, for me it works: 

  const getImageUrlBackground = (name: string) => {
    return new URL(`../assets/imgs/${name}`, import.meta.url).href;
  } 

      <NuxtImg
        :src="getImageUrlBackground('bg.webp')" 
        alt="Couple"
        as="image"
        provider="ipx"
        class="absolute inset-0 w-full h-full object-cover object-[80%_20%] md:object-center -z-10"
        preload
        sizes="100vw"
        width="1860"
        height="920"
        fetchpriority="high" 
      />

I tried the solutions on this topic but sometimes it only worked in dev, sometimes only in prod. On my mac works in dev and my images work in my deploy on netlify :D

**Comment by AndreasYNY** [comment]
+1 i'm back to using native img, using provider was not working for me either...

**Comment by AnyFork** [comment]
Does anyone have a solution to this problem?

**Comment by Smef** [comment]
I don't think this is actually an issue. While it may be confusing, this is actually expected behavior for the reasons described earlier. It would be nice if it could use import paths, but the behavior and results are expected and not a bug.

**Comment by MrEcco** [comment]
This comment https://github.com/nuxt/image/issues/1006#issuecomment-2207249013 clarifies things almost completely. Roughly, how to reproduce:

```vue
<script setup lang="ts">
const state = ref(false)
</script>

<template>
  <div
    v-if="state"
  >
    <NuxtImg
      src="/img.png"
    />
  </div>
  <button @click="() => { state = !state }">
    Click me
  </button>
</template>
```

While SSG, the builder cannot catch the request to `/_ipx/_/img.png`, so it doesn't try to render it. Unexpectedly, `preload` doesn't help here.

As a temporary workaround, I found how to fix that. You can add this "missing" route to generation via `nuxt.config.ts`:

```ts
export default defineNuxtConfig({
  nitro: {
    prerender: {
      routes: [
        '/_ipx/_/img.png' // the exact route you should identify while `npm run dev`
      ],
      crawlLinks: true
    }
  },
})
```

Hope it helps to fix this bug soon, and for developers who experience the same issue.
